### PR TITLE
#20 Query Optimisations in the PHP

### PR DESCRIPTION
**Pull request by matthewhillier**
Working through the SQL queries to optimise for performance.

The general approach here is to collect more data in fewer number of queries. I'm trying to boil this down to probably 2 or 3 larger queries, which return related sets of data, then use PHP to do the heavy-lifting in terms of completing the "last-mile" to formulate the final figures we want to display.

Adding indexes to these group-by fields should help with further speed, but those should be accompanied by changes to the CREATE TABLE scripts so can be done elsewhere.

**Issue** - https://github.com/XIVStats/XIVStats/issues/20

### CODE REVIEW

File: xiv_stats.php [hunk 1]
```diff
@@ -1285,8 +1152,8 @@
             name: 'All',
             data: [
                 <?php
-                        foreach ($america_realm_pop as $value) {
-                                echo "$value,";
+                        foreach ($american_realm_array as $value) {
+                                echo (is_null($realm_count[$value]) ? 0 : $realm_count[$value]) . ",";
```
Comment (matthewhillier) [inline xiv_stats.php]: Note to self: This is ugly - encapsulate and refactor!

File: xiv_stats.php [hunk 2]
```diff
@@ -615,13 +482,13 @@
...
                           <div class="s12 m6 l6   region-stat">
-                              <div><?php echo $fmt_europe_player_count; ?></div>
+                              <div><?php echo number_format(array_sum(array_intersect_key($realm_count, array_flip($european_realm_array)))) ?></div>
                           </div>
                       </div>
                       <div class="black-text light region-subtitle">ACTIVE PLAYERS*</div>
                       <div class="row">
                           <div class="s12 m6 l6   region-stat">
-                              <div><?php echo $fmt_europe_active_player_count; ?></div>
+                              <div><?php echo number_format(array_sum(array_intersect_key($active_realm_count, array_flip($european_realm_array)))) ?></div>
```
Comment (matthewhillier) [inline xiv_stats.php]: Note to self: encapsulate this and it's used all over

File: xiv_stats.php [hunk 3]
```diff
@@ -581,13 +447,14 @@
                       <div class="black-text light region-subtitle">ALL PLAYERS</div>
                       <div class="row">
                           <div class="s12 m6 l6   region-stat">
-                              <div><?php echo $fmt_america_player_count; ?></div>
+
```
Comment (matthewhillier) [inline xiv_stats.php]: Delete blank line

File: xiv_stats.php [hunk 4]
```diff
@@ -17,121 +17,43 @@
...
-        //$row[1] = population
-
-        //If user is not in gc, mark gc as none rather than blank
-        if ($row[0] === '') { $row[0] = "None"; }
-        $gc_active_distribution[$row[0]] = $row[1];
-}
-
-
-// Race / Gender Distribution
+$american_realm_array = array("Behemoth","Brynhildr","Diabolos","Exodus","Famfrit","Hyperion",
```
Comment (matthewhillier) [inline xiv_stats.php]: Tidy-up naming once clashes removed

**Comment by matthewhillier** [comment]
I've run this against the August 2017 database and compared values to ensure everything is reporting correctly. So far, looks good.

The compile-time using the basic mysql docker containers, with no configuration changes, was 15 minutes, 49 seconds.

I think we may be able to get come further improvements out of changing the table itself, indexes and possibly changing some of the TEXT fields to VARCHAR, otherwise I think this should be good enough for the time being.

**Comment by Pricetx** [comment]
Excellent. In the evening i'll pull down your branch into my test area and do another run on my test dataset and see how it compares to before. If all looks good i'll slot it into place for the current live run, which should start executing the PHP in a few days.

**Comment by Balls0fSteel** [comment]
I had the junk data lying around and a fast AWS machine. The script dies.

`
ubuntu@ip-172-31-13-22:~/XIVStats$ php xiv_stats.php > xiv_stats.html
PHP Notice:  Undefined variable: player_count in /home/ubuntu/XIVStats/xiv_stat$
.php on line 50
PHP Notice:  Undefined variable: active_player_count in /home/ubuntu/XIVStats/x$
v_stats.php on line 52
PHP Notice:  Undefined variable: realm_count in /home/ubuntu/XIVStats/xiv_stats.
php on line 54
PHP Notice:  Undefined index: Adamantoise in /home/ubuntu/XIVStats/xiv_stats.php
 on line 54
PHP Notice:  Undefined variable: active_realm_count in /home/ubuntu/XIVStats/xiv
_stats.php on line 56
PHP Notice:  Undefined index: Adamantoise in /home/ubuntu/XIVStats/xiv_stats.php
 on line 56
PHP Notice:  Undefined variable: gc_count in /home/ubuntu/XIVStats/xiv_stats.php
 on line 58
PHP Notice:  Undefined index: Immortal Flames in /home/ubuntu/XIVStats/xiv_stats
.php on line 58
...
PHP Notice:  Undefined variable: active_gc_count in /home/ubuntu/XIVStats/xiv_st
ats.php on line 60
PHP Notice:  Undefined index: Immortal Flames in /home/ubuntu/XIVStats/xiv_stats
.php on line 60
PHP Notice:  Undefined variable: race_gender_count in /home/ubuntu/XIVStats/xiv_
stats.php on line 63
PHP Notice:  Undefined index: Au Ra in /home/ubuntu/XIVStats/xiv_stats.php on li
ne 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on l
ine 63
PHP Notice:  Undefined variable: active_race_gender_count in /home/ubuntu/XIVSta
ts/xiv_stats.php on line 65
PHP Notice:  Undefined index: Au Ra in /home/ubuntu/XIVStats/xiv_stats.php on li
ne 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on l
ine 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on lin
e 63
...
PHP Notice:  Undefined index: Valefor in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Valefor in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Yojimbo in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Yojimbo in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zalera in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zalera in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zeromus in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zeromus in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zodiark in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zodiark in /home/ubuntu/XIVStats/xiv_stats.php on line 56
`

Hope that helps... The speed looks super fast though.

**Comment by Balls0fSteel** [comment]
I will do a second run (since it's that fast) and copy out the errors full size (tmux broke them.)

**Comment by Balls0fSteel** [comment]
`ubuntu@ip-172-31-13-22:~/XIVStats$ php xiv_stats.php > xiv_stats.html
PHP Notice:  Undefined variable: player_count in /home/ubuntu/XIVStats/xiv_stats.php on line 50
PHP Notice:  Undefined variable: active_player_count in /home/ubuntu/XIVStats/xiv_stats.php on line 52
PHP Notice:  Undefined variable: realm_count in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Adamantoise in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined variable: active_realm_count in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Adamantoise in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined variable: gc_count in /home/ubuntu/XIVStats/xiv_stats.php on line 58
PHP Notice:  Undefined index: Immortal Flames in /home/ubuntu/XIVStats/xiv_stats.php on line 58
PHP Notice:  Undefined variable: active_gc_count in /home/ubuntu/XIVStats/xiv_stats.php on line 60
PHP Notice:  Undefined index: Immortal Flames in /home/ubuntu/XIVStats/xiv_stats.php on line 60
PHP Notice:  Undefined variable: race_gender_count in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Au Ra in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined variable: active_race_gender_count in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Au Ra in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Elezen in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Elezen in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Hyur in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Hyur in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Lalafell in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Lalafell in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Miqo'te in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Miqo'te in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Roegadyn in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: Roegadyn in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: female in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 63
PHP Notice:  Undefined index: male in /home/ubuntu/XIVStats/xiv_stats.php on line 65
PHP Notice:  Undefined index: Maelstrom in /home/ubuntu/XIVStats/xiv_stats.php on line 58
PHP Notice:  Undefined index: Maelstrom in /home/ubuntu/XIVStats/xiv_stats.php on line 60
PHP Notice:  Undefined index: none in /home/ubuntu/XIVStats/xiv_stats.php on line 58
PHP Notice:  Undefined index: none in /home/ubuntu/XIVStats/xiv_stats.php on line 60
PHP Notice:  Undefined index: Order of the Twin Adder in /home/ubuntu/XIVStats/xiv_stats.php on line 58
PHP Notice:  Undefined index: Order of the Twin Adder in /home/ubuntu/XIVStats/xiv_stats.php on line 60
PHP Notice:  Undefined index: Aegis in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Aegis in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Alexander in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Alexander in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Anima in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Anima in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Asura in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Asura in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Atomos in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Atomos in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Bahamut in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Bahamut in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Balmung in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Balmung in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Behemoth in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Behemoth in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Belias in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Belias in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Brynhildr in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Brynhildr in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Cactuar in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Cactuar in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Carbuncle in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Carbuncle in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Cerberus in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Cerberus in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Chocobo in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Chocobo in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Coeurl in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Coeurl in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Diabolos in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Diabolos in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Durandal in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Durandal in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Excalibur in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Excalibur in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Exodus in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Exodus in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Faerie in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Faerie in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Famfrit in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Famfrit in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Fenrir in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Fenrir in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Garuda in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Garuda in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Gilgamesh in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Gilgamesh in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Goblin in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Goblin in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Gungnir in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Gungnir in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Hades in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Hades in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Hyperion in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Hyperion in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ifrit in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ifrit in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ixion in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ixion in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Jenova in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Jenova in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Kujata in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Kujata in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Lamia in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Lamia in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Leviathan in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Leviathan in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Lich in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Lich in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Louisoix in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Louisoix in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Malboro in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Malboro in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Mandragora in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Mandragora in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Masamune in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Masamune in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Mateus in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Mateus in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Midgardsormr in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Midgardsormr in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Moogle in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Moogle in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Odin in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Odin in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Omega in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Omega in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Pandaemonium in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Pandaemonium in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Phoenix in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Phoenix in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ragnarok in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ragnarok in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ramuh in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ramuh in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ridill in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ridill in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Sargatanas in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Sargatanas in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Shinryu in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Shinryu in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Shiva in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Shiva in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Siren in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Siren in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Tiamat in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Tiamat in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Titan in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Titan in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Tonberry in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Tonberry in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Typhon in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Typhon in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ultima in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ultima in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Ultros in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Ultros in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Unicorn in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Unicorn in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Valefor in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Valefor in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Yojimbo in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Yojimbo in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zalera in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zalera in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zeromus in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zeromus in /home/ubuntu/XIVStats/xiv_stats.php on line 56
PHP Notice:  Undefined index: Zodiark in /home/ubuntu/XIVStats/xiv_stats.php on line 54
PHP Notice:  Undefined index: Zodiark in /home/ubuntu/XIVStats/xiv_stats.php on line 56
`

Hope that helps.

**Comment by matthewhillier** [comment]
Just out of interest, was that using the table from the 1.4 gatherer? It won't work "out of the box" with older data, as there's new columns added to the DB for Stormblood support.

**Comment by Balls0fSteel** [comment]
@matthewhillier Yes, 1.4 data, the latest that the current on-going official generation also uses.
It's about 7-8 million IDs IIRC, from when we tested performance. (Before official run started.)

But yeah, **yes.**

**Comment by Balls0fSteel** [comment]
Maybe JetBrains's "PhpStorm" could identify and debug the issue? Just an idea.

**Comment by matthewhillier** [comment]
Thanks, @Balls0fSteel. Which version of php are you using? And do you still get the HTML output?

Looks like from [this post](https://stackoverflow.com/questions/39791263/how-to-fix-notice-undefined-index) that PHP is expecting some existence checking to be done, although I'm not clear if this is considered a blocking error or simply a warning statement.

I've been using the latest php docker container, which I think is >= PHP 7.1

**Comment by Balls0fSteel** [comment]
@matthewhillier 

I will have to start up a new VM but I will check if there is an HTML output (it was AWS I shot it down). 

PHP / every version is plain and latest (so fresh install + upgrade) Ubuntu 16.04, nothing else.

**Comment by Balls0fSteel** [comment]
It's
```
PHP 7.0.22-0ubuntu0.16.04.1 (cli) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.0.22-0ubuntu0.16.04.1, Copyright (c) 1999-2017, by Zend Technologies
```

By the way.

**Comment by matthewhillier** [comment]
As far as I can tell, `NOTICE` level output in PHP is basically advisory (analogous to javac warnings). They _shouldn't_ be fatal, but rather things that could cause issues. In this case, these are variables that haven't been given an initial value, and are being set by something that could set them to `null`.

Reference: [here](http://php.net/manual/en/errorfunc.configuration.php)

By default, this would be enabled as it's a development tool to help with catching potential bugs, but it's typically configured off in a production instance (& the docker container, so it seems).

I think as long as we're getting the html out consistently, it probably means we can error-handle better in the php, but shouldn't cause issues unless the sql doesn't return (in which case we're screwed anyway).

**Comment by Balls0fSteel** [comment]
So there is an HTML but some things missing here and there. I will install PHPStorm and try to look into things.

**Comment by Balls0fSteel** [comment]
Welp, PhpStorm only complained that the variables in the loop are not initialized beforehand. So I created them beforehand. Then it cried about I cannot use a "non scalar value" something. God I am bad with PHP. (Don't we want an ASP.net output page? 👍 )

**Comment by matthewhillier** [comment]
So I tried this again using the `ubuntu:latest` container and installing php fresh and I now see the PHP Notices referenced by @Balls0fSteel.

The HTML does output, and I only have gaps where there's Stormblood data missing (because it's the pre-SB data).

[index.zip](https://github.com/XIVStats/XIVStats/files/1299557/index.zip)

@Balls0fSteel - Could you attach the HTML that was created on your end so that I can see which data points didn't populate for you? Cheers!

**Comment by matthewhillier** [comment]
I've updated this branch to address the PHP Notices showing up in the debugging logging from the php process. Now showing a quiet run.

**Comment by Pricetx** [comment]
As a further note, this new version has been pulled in to the live run ahead of processing. On a dataset of 1 million characters it completed in mere seconds. It should offer a nice speed improvement.

**Pull request merged**